### PR TITLE
Scope hoisting: conditional reference of an imported, interopDefault symbol from another bundle in a case clause

### PR DESCRIPTION
**Pull request by wbinnssmith**
Similar to #4487, conditionally referencing an imported symbol originating from another bundle causes the binding to be inserted at the first reference.

This seems to only occur when the conditional reference is in a `switch` case clause that is not a block.

This PR also makes `binding` and `parent` invariant rather than conditionally used.

To do:
* [x] Implement a fix

Test Plan: Added test

### CODE REVIEW

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 6.32s | -117.00ms |
| Cached | 2.33s | +56.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 223.00ms | -41.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 222.00ms | -45.00ms 🚀 |
| dist/modern/kitchen-sink.90de237b.js | 1.16kb | +0.00b | 542.00ms | -988.00ms 🚀 |
| dist/legacy/index.html | 709.00b | +0.00b | 1.38s | +714.00ms ⚠️ |
| dist/modern/index.html | 709.00b | +0.00b | 1.38s | +713.00ms ⚠️ |
| dist/modern/styles.cda8c91b.css | 78.00b | +0.00b | 1.27s | -134.00ms 🚀 |

#### Cached Bundles

*No bundle changes detected.*

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 9.32s | +7.00ms |
| Cached | 2.22s | -81.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/main/index.js | 66.80kb | +0.00b | 285.00ms | +58.00ms ⚠️ |
| dist/module/index.js | 37.47kb | +0.00b | 283.00ms | +56.00ms ⚠️ |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.12m | +12.81s ⚠️ |
| Cached | 5.18s | -102.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/ak-editor.b5dfce6d.js | 2.50mb | -2.53kb 🚀 | 1.86m | +11.37s ⚠️ |
| dist/pdfRenderer.1ab46fe6.js | 1.11mb | +0.00b | 55.66s | -7.25s 🚀 |
| dist/editorView.2ab50f4a.js | 586.97kb | +0.00b | 1.46m | -10.51s 🚀 |
| dist/popup.c625c612.js | 218.75kb | +0.00b | 1.86m | +42.33s ⚠️ |
| dist/ak-editor.aad0c0f8.js | 136.43kb | -265.00b 🚀 | 18.96s | +2.04s ⚠️ |
| dist/Toolbar.f4bba772.js | 100.68kb | +0.00b | 58.98s | -11.43s 🚀 |
| dist/EmojiPickerComponent.d8e98777.js | 98.99kb | +0.00b | 12.37s | -4.55s 🚀 |
| dist/media-viewer.8f3d65a3.js | 71.14kb | +0.00b | 18.40s | -2.88s 🚀 |
| dist/card.78abb12a.js | 66.75kb | +0.00b | 18.37s | -2.89s 🚀 |
| dist/component.ceb96931.js | 55.90kb | +0.00b | 6.35s | -632.00ms 🚀 |
| dist/Modal.4e0c06f1.js | 41.63kb | +0.00b | 6.33s | +2.48s ⚠️ |
| dist/component.f6be2d97.js | 29.26kb | +0.00b | 8.71s | +1.66s ⚠️ |
| dist/DatePicker.ccfecb85.js | 21.58kb | +0.00b | 10.65s | -2.17s 🚀 |
| dist/EmojiPickerComponent.e0013fec.js | 18.58kb | +0.00b | 12.62s | -40.16s 🚀 |
| dist/clipboard.e8fae429.js | 17.18kb | +0.00b | 56.36s | -43.83s 🚀 |
| dist/js.baeda348.js | 16.46kb | +0.00b | 6.29s | -621.00ms 🚀 |
| dist/ui.587b7b28.js | 14.78kb | +0.00b | 17.56s | -2.08s 🚀 |
| dist/dropzone.a4525912.js | 14.70kb | +0.00b | 56.36s | -43.83s 🚀 |
| dist/smartMediaEditor.5ca64a74.js | 14.64kb | +0.00b | 57.66s | -42.53s 🚀 |
| dist/workerHasher.ebf0b386.js | 11.83kb | +0.00b | 19.08s | -33.69s 🚀 |
| dist/EmojiTypeAheadComponent.b1d18fae.js | 9.28kb | +0.00b | 1.04m | +44.09s ⚠️ |
| dist/media-viewer.52d4959e.js | 4.18kb | +0.00b | 1.09m | +31.88s ⚠️ |
| dist/EmojiPickerComponent.05ff4eeb.js | 3.92kb | +0.00b | 14.66s | -2.46s 🚀 |
| dist/16.cb2f1a3a.js | 3.72kb | +0.00b | 2.39s | -914.00ms 🚀 |
| dist/Modal.5955c887.js | 3.50kb | +0.00b | 2.39s | -4.59s 🚀 |
| dist/png-chunks-extract.c0f57e67.js | 3.49kb | +0.00b | 6.29s | -621.00ms 🚀 |
| dist/dropzone.79e05ece.js | 3.44kb | +0.00b | 1.09m | -3.41s 🚀 |
| dist/ak-editor.47799eae.css | 3.43kb | +0.00b | 6.60s | -1.08m 🚀 |
| dist/16.c7be7376.js | 2.54kb | +0.00b | 6.47s | -442.00ms 🚀 |
| dist/EmojiUploadComponent.7cd6a990.js | 2.50kb | +0.00b | 14.66s | -3.41s 🚀 |
| dist/ResourcedEmojiComponent.8ea5609c.js | 2.29kb | +0.00b | 8.66s | +1.69s ⚠️ |
| dist/card.c002a0cc.js | 2.19kb | +0.00b | 32.71s | -6.69s 🚀 |
| dist/media-picker-analytics-error-boundary.bd391930.js | 1.84kb | +0.00b | 55.66s | -13.47s 🚀 |
| dist/browser.1bc939d3.js | 1.81kb | +0.00b | 57.52s | -42.66s 🚀 |
| dist/date.eb45fb3a.js | 1.81kb | +0.00b | 7.02s | -406.00ms 🚀 |
| dist/16.b1b0372d.js | 1.77kb | +0.00b | 6.56s | -530.00ms 🚀 |
| dist/feedback.d3363837.js | 1.71kb | +0.00b | 10.08s | -2.33s 🚀 |
| dist/workerHasher.ac36166a.js | 1.64kb | +0.00b | 14.90s | -37.87s 🚀 |
| dist/ak-editor.41b1867d.js | 1.60kb | +0.00b | 58.98s | -11.59s 🚀 |
| dist/list-number.d99b1635.js | 1.53kb | +0.00b | 7.24s | -429.00ms 🚀 |
| dist/status.9fcfe9a8.js | 1.53kb | +0.00b | 7.45s | -391.00ms 🚀 |
| dist/16.450f3a0d.js | 1.49kb | +0.00b | 6.71s | -2.25s 🚀 |
| dist/code.ace3b8b0.js | 1.45kb | +0.00b | 7.02s | -351.00ms |
| dist/16.2b801dd1.js | 1.44kb | +0.00b | 6.48s | -610.00ms 🚀 |
| dist/16.fe51c023.js | 1.44kb | +0.00b | 6.69s | -2.28s 🚀 |
| dist/16.6b0e338c.js | 1.38kb | +0.00b | 6.92s | -351.00ms 🚀 |
| dist/link.6d7c2845.js | 1.38kb | +0.00b | 7.18s | -396.00ms 🚀 |
| dist/heading6.fec11280.js | 1.38kb | +0.00b | 10.07s | +1.80s ⚠️ |
| dist/16.e32aa49c.js | 1.34kb | +0.00b | 2.39s | -895.00ms 🚀 |
| dist/16.50f9c790.js | 1.33kb | +0.00b | 6.88s | -351.00ms 🚀 |
| dist/media-card-analytics-error-boundary.65fd9d6a.js | 1.30kb | +0.00b | 33.08s | -29.84s 🚀 |
| dist/emoji.d686023a.js | 1.30kb | +0.00b | 7.14s | -377.00ms 🚀 |
| dist/16.d1b58d7c.js | 1.29kb | +0.00b | 7.01s | -364.00ms 🚀 |
| dist/16.1dbb5df4.js | 1.28kb | +0.00b | 2.39s | -894.00ms 🚀 |
| dist/Modal.6509cbe3.js | 1.26kb | +0.00b | 6.29s | -687.00ms 🚀 |
| dist/heading5.9d7bec75.js | 1.25kb | +0.00b | 7.59s | -670.00ms 🚀 |
| dist/16.ceac57cd.js | 1.25kb | +0.00b | 6.77s | -371.00ms 🚀 |
| dist/16.0efadc10.js | 1.23kb | +0.00b | 6.56s | -536.00ms 🚀 |
| dist/16.35be54a4.js | 1.23kb | +0.00b | 6.71s | -2.25s 🚀 |
| dist/16.6be9bbc8.js | 1.23kb | +0.00b | 6.77s | -366.00ms 🚀 |
| dist/16.9015e8cf.js | 1.23kb | +0.00b | 6.48s | -503.00ms 🚀 |
| dist/16.234ee133.js | 1.23kb | +0.00b | 2.39s | -895.00ms 🚀 |
| dist/expand.c08c17cb.js | 1.23kb | +0.00b | 10.65s | -2.17s 🚀 |
| dist/layout.a74cdd85.js | 1.12kb | +0.00b | 7.17s | -412.00ms 🚀 |
| dist/divider.dcf70baf.js | 1.10kb | +0.00b | 7.10s | -413.00ms 🚀 |
| dist/quote.4c153635.js | 1.09kb | +0.00b | 7.45s | -391.00ms 🚀 |
| dist/action.862d8d7d.js | 1.08kb | +0.00b | 7.02s | -351.00ms |
| dist/decision.15d33659.js | 1.06kb | +0.00b | 7.04s | -397.00ms 🚀 |
| dist/panel-warning.36e3f1a9.js | 1.06kb | +0.00b | 7.39s | -374.00ms 🚀 |
| dist/list.c832f8d2.js | 1.03kb | +0.00b | 7.26s | -367.00ms 🚀 |
| dist/panel-error.0cf404ec.js | 981.00b | +0.00b | 7.31s | -398.00ms 🚀 |
| dist/panel.32d95efd.js | 972.00b | +0.00b | 7.40s | -399.00ms 🚀 |
| dist/table.92481ac8.js | 964.00b | +0.00b | 7.45s | -422.00ms 🚀 |
| dist/panel-success.c8bee749.js | 920.00b | +0.00b | 12.13s | +4.37s ⚠️ |
| dist/panel-note.c8014b86.js | 916.00b | +0.00b | 7.32s | -387.00ms 🚀 |
| dist/EmojiTypeAheadComponent.441c229c.js | 619.00b | +0.00b | 16.54s | -3.06s 🚀 |
| dist/simpleHasher.dd9d4306.js | 609.00b | +0.00b | 12.57s | -40.19s 🚀 |
| dist/index.html | 219.00b | +0.00b | 7.18s | +6.34s ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/ak-editor.b5dfce6d.js | 2.50mb | -2.53kb 🚀 | 629.00ms | -11.00ms |
| dist/pdfRenderer.1ab46fe6.js | 1.11mb | -4.00b 🚀 | 625.00ms | -10.00ms |
| dist/ak-editor.aad0c0f8.js | 136.43kb | -265.00b 🚀 | 626.00ms | -11.00ms |

</p></details>

<details><summary>Three.js x4 🚨</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | **FAILED** | -0.00ms |
| Cached | **FAILED** | -0.00ms |

#### Cold Bundles

*No bundles found, this is probably a failed build...*
#### Cached Bundles

*No bundles found, this is probably a failed build...*
</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/264840925233021450)**

File: packages/shared/scope-hoisting/src/link.js [hunk 1]
```diff
@@ -217,35 +217,33 @@ export function link({
     if (mod.meta.isCommonJS && originalName === 'default') {
       let name = getName(mod, '$interop$default');
       if (!path.scope.getBinding(name)) {
-        // Hoist to the nearest path with the same scope as the exports is declared in
-        let binding = path.scope.getBinding(
-          assertString(mod.meta.exportsIdentifier),
+        let binding = nullthrows(
```
Comment (wbinnssmith) [inline packages/shared/scope-hoisting/src/link.js:220]: @mischnic I believe this is a same assumption to make. Let me know if you know of a case where the binding may not exist.

File: packages/shared/scope-hoisting/src/link.js [hunk 2]
```diff
@@ -217,35 +217,33 @@ export function link({
...
+          path.scope.getBinding(
+            bundle.hasAsset(mod)
+              ? assertString(mod.meta.exportsIdentifier)
+              : // If this bundle doesn't have the asset, use the binding for
+                // the `parcelRequire`d init function.
+                getIdentifier(mod, 'init').name,
```
Comment (mischnic) [inline packages/shared/scope-hoisting/src/link.js:226]:
```suggestion
                getName(mod, 'init'),
```
The identifier is actually already in the `node` parameter, but it's either an Identifier or a CallExpression. Not really cleaner.

File: packages/shared/scope-hoisting/src/link.js (hunk 1 above)
Comment (mischnic) [reply packages/shared/scope-hoisting/src/link.js:220]: Yes, in the `hasAsset` case, it obviously is. And `addBundleImport` creates a binding:
https://github.com/parcel-bundler/parcel/blob/52f0cbbde1ffc85986781c8590b5cf19fcc34e7d/packages/shared/scope-hoisting/src/link.js#L329-L336

**Pull request merged**